<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="dual_arm_hong_kong">

  <!-- common stuff -->
  <xacro:include filename="$(find ur_description)/urdf/common.gazebo.xacro" />

  <!-- s-model robotiq gripper -->
  <xacro:include filename="$(find robotiq_s_model_visualization)/cfg/s-model_articulated_macro.xacro" />

  <!-- arg2f model robotiq gripper -->
  <xacro:include filename="$(find robotiq_arg2f_model_visualization)/urdf/robotiq_arg2f_140_model_macro.xacro" />

  <xacro:include filename="$(find ur_description)/urdf/ur10.urdf.xacro" />

  <xacro:ur10_robot prefix="hong_" joint_limited="true"
     shoulder_pan_lower_limit="${-pi}" shoulder_pan_upper_limit="${pi}"
     shoulder_lift_lower_limit="${-pi}" shoulder_lift_upper_limit="${pi}"
     elbow_joint_lower_limit="${-pi}" elbow_joint_upper_limit="${pi}"
     wrist_1_lower_limit="${-pi}" wrist_1_upper_limit="${pi}"
     wrist_2_lower_limit="${-pi}" wrist_2_upper_limit="${pi}"
     wrist_3_lower_limit="${-pi}" wrist_3_upper_limit="${pi}"/>

  <xacro:ur10_robot prefix="kong_" joint_limited="true"
     shoulder_pan_lower_limit="${-pi}" shoulder_pan_upper_limit="${pi}"
     shoulder_lift_lower_limit="${-pi}" shoulder_lift_upper_limit="${pi}"
     elbow_joint_lower_limit="${-pi}" elbow_joint_upper_limit="${pi}"
     wrist_1_lower_limit="${-pi}" wrist_1_upper_limit="${pi}"
     wrist_2_lower_limit="${-pi}" wrist_2_upper_limit="${pi}"
     wrist_3_lower_limit="${-pi}" wrist_3_upper_limit="${pi}"/>


  <xacro:robotiq_arg2f_140 prefix="two_fingered_"/>

  <!-- start of gripper force-torque sensor (coupler) -->
  <link name="kong_coupler">
    <visual>
      <geometry>
        <cylinder length="0.035" radius="0.03504"/>
      </geometry>
      <material name="grey">
        <color rgba="0.7 0.7 0.7 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.035" radius="0.03504"/>
      </geometry>
      <material name="yellow">
        <color rgba="0 1 1 1"/>
      </material>
    </collision>
  </link>
  <!-- end of gripper force-torque sensor (coupler) -->

  <!-- start of arm-coupler joint definitions -->
  <joint name="kong_wrist_to_coupler" type="fixed">
    <parent link="kong_ee_link"/>
    <child link="kong_coupler"/>
    <origin xyz="0.0175 0 0" rpy="0 ${pi/2} 0"/>
  </joint>
  <!-- end of arm-coupler joint definitions -->

  <!-- start of coupler-gripper joint definitions -->
  <joint name="kong_coupler_to_gripper" type="fixed">
    <parent link="kong_coupler"/>
    <child link="two_fingered_robotiq_arg2f_base_link"/>
    <origin xyz="0 0 0.015" rpy="0 0 0"/>
  </joint>
  <!-- end of coupler-gripper coupling joint definitions -->

  <xacro:s-model_articulated prefix="three_fingered_"/>

  <link name="hong_coupler">
    <visual>
      <geometry>
        <cylinder length="0.035" radius="0.03504"/>
      </geometry>
      <material name="grey">
        <color rgba="0.7 0.7 0.7 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.035" radius="0.03504"/>
      </geometry>
      <material name="yellow">
        <color rgba="0 1 1 1"/>
      </material>
    </collision>
  </link>
  <!-- end of gripper force-torque sensor (coupler) -->

  <!-- start of arm-coupler joint definitions -->
  <joint name="hong_wrist_to_coupler" type="fixed">
    <parent link="hong_ee_link"/>
    <child link="hong_coupler"/>
    <origin xyz="0.0175 0 0" rpy="0 ${pi/2} 0"/>
  </joint>
  <!-- end of arm-coupler joint definitions -->

  <!-- start of coupler-gripper joint definitions -->
  <joint name="hong_coupler_to_gripper" type="fixed">
    <parent link="hong_coupler"/>
    <child link="three_fingered_palm"/>
    <origin xyz="0 0 0.065" rpy="${pi/2} 0 ${-pi/2}"/>
  </joint>
  <!-- end of coupler-gripper coupling joint definitions -->


  <!-- Table over which the two arms are mounted/-->
  <link name="table">
    <visual>
      <geometry>
        <box size="1.80 2.40 0.2"/>
      </geometry>
    </visual>
  </link>


  <joint name="table_to_hong" type="fixed">
    <parent link="table" />
    <child link = "hong_base_link" />
    <origin xyz="-0.64 0.94 0.1" rpy="0.0 0.0 ${pi}" />
  </joint>


  <joint name="table_to_kong" type="fixed">
    <parent link="table" />
    <child link = "kong_base_link" />
    <origin xyz="-0.64 -0.94 0.1" rpy="0.0 0.0 0.0" />
  </joint>


  <link name="world" />

  <joint name="world_joint" type="fixed"> <!--World frame coincides with kong_base_link-->
    <parent link="world" />
    <child link = "table" />
    <origin xyz="0.64 0.94 -0.10" rpy="0.0 0.0 0.0" />
  </joint>



</robot>
